libleptongui/scheme/schematic/builtins.scm

```diff
@@ -861,14 +862,209 @@ the snap grid size should be set to 100")))
 ;; -------------------------------------------------------------------
 ;;;; Hierarchy actions
 
-(define-action-public (&hierarchy-down-schematic #:label (G_ "Down Schematic") #:icon "gtk-go-down")
-  (run-callback i_callback_hierarchy_down_schematic "&hierarchy-down-schematic"))
+(define (failed-to-descend-error filename message)
+  (log! 'message (G_ "Failed to descend into ~S: ~A") filename message)
+  (format #f (string-append
+              (format #f
+                      (G_ "Failed to descend hierarchy into ~S: ~A\n\n")
+                      filename
+                      message)
+              (G_ "The lepton-schematic log may contain more information."))))
+
+(define (source-attrib? attrib)
+  (string= (attrib-name attrib) "source"))
+
+(define (split-attrib-value object)
+  (string-split (attrib-value object) #\,))
+
+(define (zoom-child-page *window *parent *child)
+  (define *toplevel (gschem_toplevel_get_toplevel *window))
+  (define *page-view (gschem_toplevel_get_current_page_view *window))
+  (lepton_toplevel_goto_page *toplevel *child)
+  (gschem_toplevel_page_changed *window)
+  (gschem_page_view_zoom_extents *page-view %null-pointer)
+  (undo-save-state)
+  (lepton_toplevel_goto_page *toplevel *parent)
+  (gschem_toplevel_page_changed *window))
+
+(define (hierarchy-down-error-dialog filename *error)
+  (let ((secondary-message
+         (failed-to-descend-error filename
+                                  (if (null-pointer? *error)
+                                      (G_ "Unknown error.")
+                                      (gerror-message (dereference-pointer *error))))))
+
+    (generic_error_dialog (string->pointer (G_ "Failed to descend hierarchy."))
+                          (string->pointer secondary-message))
+
+    (g_clear_error *error)))
+
+(define (hierarchy-down-filename filename *window *parent page-control)
+  (define use-tabs? (true? (x_tabs_enabled)))
+
+  (log! 'message (G_ "Searching for source ~S") filename)
+  (let* ((*error (bytevector->pointer (make-bytevector (sizeof '*) 0)))
+         (*child (s_hierarchy_down_schematic_single *window
```

Comment:
If `s_hierarchy_down_schematic_single()` returns NULL here,
you eventually get a "dereference null pointer" exception.
`hierarchy-down-error-dialog()` -> `gerror-message()`:
because bytevector(0) is not a %null-pointer.